### ISSUE TITLE
"pbench-sysstat-12.0.3 is not installed" showing in metadata.log after running pbench-user-benchmark

### ISSUE THREAD

**Issue by sanjaychari** [distributed-system-analysis/pbench#2289]
With the latest build of pbench-agent https://copr.fedorainfracloud.org/coprs/portante/pbench-test/build/2162023/, after running pbench-user-benchmark, metadata.log shows `pbench-sysstat-12.0.3 is not installed`, even though it is installed. The logs are below.
```
pbench-user-benchmark -- sleep 60
pbench-tool-data-sink: connected to redis server localhost:17001
pbench-tool-meister: connected to redis server localhost:17001
Running sleep 60 for iteration 1-default
iostat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/iostat', missing
iostat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/iostat', missing
mpstat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/mpstat', missing
iostat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/iostat', missing
mpstat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/mpstat', missing
perf: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/perf', missing
iostat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/iostat', missing
mpstat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/mpstat', missing
perf: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/perf', missing
pidstat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/pidstat', missing
iostat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/iostat', missing
mpstat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/mpstat', missing
perf: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/perf', missing
pidstat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/pidstat', missing
proc-interrupts: post-processing data
proc-vmstat: post-processing data
sar: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/sar', missing
Collecting system information
```
```
dnf install -y pbench-sysstat
Updating Subscription Management repositories.
Last metadata expiration check: 0:01:04 ago on Thu 06 May 2021 12:17:45 PM UTC.
Package pbench-sysstat-12.0.3-1.x86_64 is already installed.
Dependencies resolved.
Nothing to do.
Complete!
```
```
cat /var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/metadata.log
[pbench]
config = 
date = 2021-05-06T12:19:04
name = pbench-user-benchmark__2021.05.06T12.19.04
rpm-version = v0.71.0-171gfe0675763
script = pbench-user-benchmark
iterations = 1-default

[controller]
hostname = ip-172-31-94-3.ec2.internal
hostname-s = ip-172-31-94-3
hostname-f = ip-172-31-94-3.ec2.internal
hostname-i = 172.31.94.3
hostname-a = ip-172-31-94-3.ec2.internal
ssh_opts = -o StrictHostKeyChecking=no

[run]
controller = ip-172-31-94-3.ec2.internal
start_run = 2021-05-06T12:19:05.489913
end_run = 2021-05-06T12:20:25.934423

[tools]
hosts = ip-172-31-94-3.ec2.internal
group = default
trigger = None

[tools/ip-172-31-94-3.ec2.internal]
label = 
tools = iostat,mpstat,perf,pidstat,proc-interrupts,proc-vmstat,sar,turbostat
hostname-s = ip-172-31-94-3
hostname-f = ip-172-31-94-3.ec2.internal
hostname-i = 172.31.94.3
hostname-a = ip-172-31-94-3.ec2.internal
rpm-version = v0.71.0-171gfe0675763
iostat = --interval=3
mpstat = --interval=3
perf = --record-opts='-a --freq=100'
pidstat = --interval=30
proc-interrupts = --interval=3
proc-vmstat = --interval=3
sar = --interval=3
turbostat = --interval=3

[tools/ip-172-31-94-3.ec2.internal/iostat]
options = --interval=3
install_check_status_code = 1
install_check_output = iostat: pbench-sysstat-12.0.3 is not installed

[tools/ip-172-31-94-3.ec2.internal/mpstat]
options = --interval=3
install_check_status_code = 1
install_check_output = mpstat: pbench-sysstat-12.0.3 is not installed

[tools/ip-172-31-94-3.ec2.internal/perf]
options = --record-opts='-a --freq=100'
install_check_status_code = 1
install_check_output = perf: perf is not installed

[tools/ip-172-31-94-3.ec2.internal/pidstat]
options = --interval=30
install_check_status_code = 1
install_check_output = pidstat: pbench-sysstat-12.0.3 is not installed

[tools/ip-172-31-94-3.ec2.internal/proc-interrupts]
options = --interval=3
install_check_status_code = 0
install_check_output = 

[tools/ip-172-31-94-3.ec2.internal/proc-vmstat]
options = --interval=3
install_check_status_code = 0
install_check_output = 

[tools/ip-172-31-94-3.ec2.internal/sar]
options = --interval=3
install_check_status_code = 1
install_check_output = sar: pbench-sysstat-12.0.3 is not installed

[tools/ip-172-31-94-3.ec2.internal/turbostat]
options = --interval=3
install_check_status_code = 0
install_check_output = turbostat: kernel-tools is installed

[iterations/1-default]
iteration_number = 1
iteration_name = 1-default
user_script = sleep, 60
```

**Comment by sanjaychari** [comment]
Ignore the line `perf is not installed` in metadata.log. That is because I forgot to run `dnf install perf`, and is not related to this issue. This issue is only related to pbench-sysstat.

**Comment by ndokos** [comment]
I can reproduce this. I'll add an update once I've figured out where it's going wrong.

**Comment by ndokos** [comment]
The call to `/opt/pbench-agent/tool-scripts/iostat --install` fails. I believe this was introduced with 5794fc1db134543d52637b3903891157a34bc77c

Running the above with `-x` shows:

```
...
+ require_rpm pbench-sysstat 12.0.3
+ _tpn=pbench-sysstat
+ _tpv=12.0.3
++ require-rpm pbench-sysstat 12.0.3
+ _installed_rpm=pbench-sysstat-12.0.3
+ rc=1
+ [[ 1 != 0 ]]
...
```
I guess the recursive call to `require-rpm` messes up the return code, but I'm not sure yet.

**Comment by ndokos** [comment]
@portante: there is a `check_installed_rpm` in `base` that calls `require-rpm`, there is a `require-rpm` script in `util-scripts` and there is a (recursive!) `require-rpm` function in `tool-scripts/base-tool`. I think that's the underlying  cause of this problem. Do we need the function at all? Shouldn't we be calling `check_installed_rpm` from `base-tool`?

**Comment by ndokos** [comment]
Scratch that: the function is `require_rpm` in `base-tool` and it calls the `require-rpm` executable in `util-scripts`, but this returns status 1 when called as `require-rpm pbench-sysstat 12.0.3`.

**Comment by portante** [comment]
Working as expected.  :(

We have `12.0.3-1` installed, but we want an exact match, so `12.0.3` is not the same as `12.0.3-1`.  We need to fix the required default to be `12.0.3-1`.

**Comment by webbnh** [comment]
@ndokos, I feel your pain:  I ran into the underscore vs. hyphen confusion in the shell function vs. script during code review -- I guess I should have followed my instinct and objected.  We should "never" have two references which differ by one letter, by the case of the letters, or by any other subtlety, such as the difference between a hyphen and an underscore.  :-(
